sunshine: new Portfile

ReenigneArcher opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 12.4 21F79 x86_64
Xcode 13.4 13F17a
Command Line Tools 13.4

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`? (tried, but errors, no tests turned on)
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

ReenigneArcher commented:
I don't believe this will build on anything less than BigSur, based on this comment from the contributor who added MacOS support initially. https://github.com/SunshineStream/Sunshine/discussions/117#discussioncomment-2513494

Is that a requirement to be included in this repo?

mascguy commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+PortGroup                boost 1.0
+
+# bump revision when changes are made to this file
+revision                 1
```
You'll want to start at revision zero for a new port.

Also, in terms of best practices, it's preferable to place this immediately after the `version` line.

ReenigneArcher commented:
> Building on earlier versions isn't an absolute requirement, but it's preferable to do so when possible.
> 
> Based on the discussion, it sounds like hardware encoding won't be available with macOS versions prior to Big Sur. So if we are able to compile for earlier versions, we should also add a note mentioning that.

Okay, thanks for the quick response! I'll see what I can do to get it to compile the earlier version and report back.

pmetzger commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+license                  GPL-3
+maintainers              {@SunshineStream sunshinestream}
+description              Sunshine is a Gamestream host for Moonlight.
+long_description         {*}${description}
```
This description is inadequate. Unless someone knows what a "Gamestream host for Moonlight" is, they will have no idea what this port does. (I have no idea.) The Long Description should be sufficient for a person who is not familiar with something already to figure out whether they should be interested in it or not. Ideally it should be several sentences of English text in a case like this.

ReenigneArcher replied:
Thanks... I'll update it to give a little more context, although if someone doesn't know what gamestreaming or Moonlight is they likely will have no interest in Sunshine.

reneeotten commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+    xinstall -d -m 755 ${destroot}${prefix}/etc/${name}/config
+
+    # install sunshine.conf
+    xinstall {*}[glob ${worksrcpath}/src_assets/common/config/*.*] ${destroot}${prefix}/etc/${name}/config
```
you probably don't want to overwrite a configuration file if someone would update to a newer version. I'm sure there are many Portfiles that take care of things like that and/or there might be a description in the guide/Wiki on how to handle this. Perhaps the `phpmyadmin` is one example where you can see that it checks if a configuration file is already present and depending on the outcome does a different action.

ReenigneArcher replied:
I tried to find examples but couldn't find one.

The documentation is not complete. https://guide.macports.org/#development.practices.dont-overwrite

reneeotten replied:
I suggested a port that does similar things to check how one could do this (i.e., `phpadmin`); another one that uses `glob` is the `cppcheck` port. 

I agree that some parts of the guide are incomplete; sometimes you can find that information in the Wiki. For example, in this case you could check [here](https://trac.macports.org/wiki/PortfileRecipes#configfiles).

ReenigneArcher replied:
Thank you, sorry I must have read it too fast previously... The link you provided is good though!

I'll use these as references:
https://trac.macports.org/browser/trunk/dports/security/stegdetect/Portfile#L49
https://trac.macports.org/browser/trunk/dports/sysutils/bacula/Portfile#L210

ReenigneArcher replied:
I'm sure I did something wrong, but tested this and it doesn't appear the `post-activate` works with a `pkg` generated using `sudo port pkg sunshine`

```txt
# Rename files in `destroot`
post-destroot {
    file rename ${destroot}${prefix}/etc/${name}/config/sunshine.conf ${destroot}${prefix}/etc/${name}/config/sunshine.conf.sample
    file rename ${destroot}${prefix}/etc/${name}/config/apps.json ${destroot}${prefix}/etc/${name}/config/apps.json.sample
}

# Don't overwrite existing preference files
post-activate {
    if {![file exists ${prefix}/etc/${name}/config/apps.json]} {
        file copy ${destroot}${prefix}/etc/${name}/config/apps.json.sample \
            ${prefix}/etc/${name}/config/apps.json
        }
    if {![file exists ${prefix}/etc/${name}/config/sunshine.conf]} {
        file copy ${destroot}${prefix}/etc/${name}/config/sunshine.conf.sample \
            ${prefix}/etc/${name}/config/sunshine.conf
        }
}
```

```txt
reenignearcher@ReenigneArchers-iMac-Pro MacOS % ls /opt/local/etc/Sunshine/config
apps.json.mp_1655322422		sunshine.conf.mp_1655322422
apps.json.sample		sunshine.conf.sample
```

It does work when installing using the Portfile. Is there a way to do this that will work for the `pkg` as well?

reneeotten reviewed (changes requested):
thanks @ReenigneArcher for the new submission. A few comments after a quick read through

ReenigneArcher commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+categories               multimedia emulators games
+platforms                darwin
+license                  GPL-3
+maintainers              {@SunshineStream sunshinestream}
```
The Portfile is configured by cmake in our main repo, eventually I'll automate the update process, so I'd prefer not to have `openmaintainer`. Is an email required? I prefer users just reach out on github.

reneeotten replied:
nothing is strictly required I think, but certainly preferred. The argument about automating the update process doesn't really prohibit to add `openmaintainer`, but again it's up to you.

ReenigneArcher commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+                         port:ffmpeg \
+                         port:libopus
+
+boost.version            1.76
```
You appear to not like being explicit... In my opinion it's better to explicitly state what you want, so when someone changes the defaults, it still works...

reneeotten replied:
I am not sure what you mean with "to not like being explicit"; we have defaults set by PortGroups or `macports-base` for a reason, namely such that every individual port doesn't need to repeat that in its Portfile.

Ideally we want ports to use the specific version of boost (or some other things like Python, Perl, ...) that MacPorts considers the default and only override that when a package doesn't build with it. Typically, a change of the default versions will be announced on the "-devel" mailinglist giving maintainers time to check whether "their" ports build with that new default, and, if not, to specify a compatible version.

Having said that, if you want to specify `boost.version` that's fine, I just pointed out that it is not required since you're repeating the default.

reneeotten commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+# bump revision when changes are made to this file
+revision                 1
+
+github.setup             sunshinestream sunshine master
```
sorry, I see upstream doesn't actually provide "releases"; so you should use "archive" instead in the above snippet.

ReenigneArcher replied:
We provide releases... but the port file will be generated before a release is published/tagged. Is it possible to use a commit instead? Also, I'm working on integration building the port inside PR checks, to ensure no breaking changes so using the commit would be the preferred option if it's possible.

reneeotten replied:
that sounds like the "wrong" order from our point of view ;) Automation is great, but updating a Portfile to actually use the release is what we would want - so no, we don't want to have a commit hash (unless it's a `-devel` port).

reneeotten commented on multimedia/Sunshine/Portfile:
```diff
@@ -0,0 +1,75 @@
+revision                 1
+
+github.setup             sunshinestream sunshine master
+name                     Sunshine
```
or is you want to you should use the same capitalization as for the upstream website

ReenigneArcher replied:
> or is you want to you should use the same capitalization as for the upstream website

I'm confused by this... the repo name is capitalized in github if that's what you mean: https://github.com/SunshineStream/Sunshine

reneeotten replied:
oh, I see... sorry didn't realize that the upstream name is capitalized. In that case should also use the capitalization of the "github.author" (i.e., SunshineStream) and "github.project" (i.e., Sunshine) in the `github.setup` line.

ReenigneArcher commented:
@mascguy 

> Building on earlier versions isn't an absolute requirement, but it's preferable to do so when possible.

Haven't had much luck with this, and as Catalina will be end of life in 5 months (November 30, 2022), I probably won't put in the effort to attempt to get this to compile on 10.15 where performance would be terrible anyway. I'm guessing at that point github will also remove the 10.15 runner.

mascguy commented:
> Haven't had much luck with this, and as Catalina will be end of life in 5 months (November 30, 2022), I probably won't put in the effort to attempt to get this to compile on 10.15 where performance would be terrible anyway. I'm guessing at that point github will also remove the 10.15 runner.

That's fine, given the circumstances.

ReenigneArcher commented:
> That's fine, given the circumstances.

Thanks. I'll try to get a new file pushed soon. I've addressed most of these concerns in the source repo (not the macports fork).
